src/Elastic.Apm/Helpers/DbConnectionStringParser.cs

```diff
@@ -0,0 +1,236 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.Helpers
+{
+	internal class DbConnectionStringParser
+	{
+		private const string ThisClassName = nameof(DbConnectionStringParser);
+		internal const int MaxCacheSize = 100;
+
+		private readonly IApmLogger _logger;
+		private readonly ThreadLocal<Dictionary<string, Destination>> _cache =
```

Comment:
I am surprised by using `ThreadLocal` here. Diagnostic source listeners are executed in the thread from the thread pool, which is not the same each time. I run a small snippet to check it and got the following results (it was release configuration):

```
Thread Id: 7, Is background: True, Is from thread pool: True
Request '6' started: { "isMaster" : 1, "client" : { "driver" : { "name" : "mongo-csharp-driver", "version" : "2.9.2.0" }, "os" : { "type" : "Windows", "name" : "Microsoft Windows 10.0.18362", "architecture" : "x86_64", "version" : "10.0.18362" }, "platform" : "
.NET Core 3.0.1" }, "compression" : [] }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '6' succeeded for isMaster in 14.1601ms with { "ismaster" : true, "maxBsonObjectSize" : 16777216, "maxMessageSizeBytes" : 48000000, "maxWriteBatchSize" : 100000, "localTime" : ISODate("2019-12-20T10:36:57.418Z"), "logicalSessionTimeoutMinutes" : 30, "co
nnectionId" : 8, "minWireVersion" : 0, "maxWireVersion" : 8, "readOnly" : false, "ok" : 1.0 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '7' started: { "buildInfo" : 1 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '7' succeeded for buildInfo in 1.8988ms with { "version" : "4.2.2", "gitVersion" : "a0bbbff6ada159e19298d37946ac8dc4b497eadf", "modules" : [], "allocator" : "tcmalloc", "javascriptEngine" : "mozjs", "sysInfo" : "deprecated", "versionArray" : [4, 2, 2, 0
], "openssl" : { "running" : "OpenSSL 1.1.1  11 Sep 2018", "compiled" : "OpenSSL 1.1.1  11 Sep 2018" }, "buildEnvironment" : { "distmod" : "ubuntu1804", "distarch" : "x86_64", "cc" : "/opt/mongodbtoolchain/v3/bin/gcc: gcc (GCC) 8.2.0", "ccflags" : "-fno-omit-fr
ame-pointer -fno-strict-aliasing -ggdb -pthread -Wall -Wsign-compare -Wno-unknown-pragmas -Winvalid-pch -Werror -O2 -Wno-unused-local-typedefs -Wno-unused-function -Wno-deprecated-declarations -Wno-unused-const-variable -Wno-unused-but-set-variable -Wno-missing
-braces -fstack-protector-strong -fno-builtin-memcmp", "cxx" : "/opt/mongodbtoolchain/v3/bin/g++: g++ (GCC) 8.2.0", "cxxflags" : "-Woverloaded-virtual -Wno-maybe-uninitialized -fsized-deallocation -std=c++17", "linkflags" : "-pthread -Wl,-z,now -rdynamic -Wl,--
fatal-warnings -fstack-protector-strong -fuse-ld=gold -Wl,--build-id -Wl,--hash-style=gnu -Wl,-z,noexecstack -Wl,--warn-execstack -Wl,-z,relro", "target_arch" : "x86_64", "target_os" : "linux" }, "bits" : 64, "debug" : false, "maxBsonObjectSize" : 16777216, "st
orageEngines" : ["biggie", "devnull", "ephemeralForTest", "wiredTiger"], "ok" : 1.0 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '8' started: { "getLastError" : 1 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '8' succeeded for getLastError in 1.3274000000000001ms with { "connectionId" : 8, "n" : 0, "syncMillis" : 0, "writtenTo" : null, "err" : null, "ok" : 1.0 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '9' started: { "find" : "people", "filter" : { }, "$db" : "diagnosticsource", "lsid" : { "id" : CSUUID("654c3f97-6643-4fa5-828e-ab474f8807fa") } }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 7, Is background: True, Is from thread pool: True
Request '9' succeeded for find in 8.997200000000001ms with { "cursor" : { "firstBatch" : [{ "_id" : ObjectId("5dfca35bb523d240ef428a13"), "name" : "Joe Smith", "email" : "jsmith@gmail.com", "age" : 40, "admin" : false }, { "_id" : ObjectId("5dfca35bb523d240ef42
8a14"), "name" : "Jen Ford", "email" : "jford@gmail.com", "age" : 45, "admin" : true }], "id" : NumberLong(0), "ns" : "diagnosticsource.people" }, "ok" : 1.0 }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 7, Is background: True, Is from thread pool: True
Request '10' started: { "$db" : "diagnosticsource", "lsid" : { "id" : CSUUID("654c3f97-6643-4fa5-828e-ab474f8807fa") } }
---   ---   ---   ---   ---   ---   ---   ---   ---
Thread Id: 8, Is background: True, Is from thread pool: True
Request '10' failed with 'MongoDB.Driver.MongoCommandException: $db command failed'
---   ---   ---   ---   ---   ---   ---   ---   ---
```

So, using `ThreadLocal` is less efficient for a caching scenario in our use case than `ConcurrentDictionary`. What was your motivation?

Reply:
Just chiming in quickly: I haven't look at the whole PR yet, just parsed it a little bit last evening and I also had the same question... I also thought that I'd just use `static ConcurrentDictionary`, unless there is a good reason not to.

Reply:
It's an interesting point - I'll compare performance characteristics of using `ThreadLocal` vs. `ConcurrentDictionary`.

Reply:
Done